pkg/util/client/client.go

```diff
@@ -31,23 +31,45 @@ func CreatePatch(before, after client.Object) (client.Patch, error) {
 	return client.RawPatch(patchBase.Type(), patchBytes), nil
 }
 
-func Patch(ctx context.Context, c client.Client, before, after client.Object) error {
-	patch, err := CreatePatch(before, after)
+// Patch applies the merge patch of client.Object with SSA.
+// If strict is true, the resourceVersion will be part of the patch, make this call fail if
+// client.Object was changed.
+func Patch(ctx context.Context, c client.Client, obj client.Object, strict bool, update func() (bool, error)) error {
+	objOriginal := obj.DeepCopyObject().(client.Object)
+	if strict {
+		objOriginal.SetResourceVersion("")
+	}
+	updated, err := update()
+	if err != nil || !updated {
+		return err
+	}
+	patch, err := CreatePatch(objOriginal, obj)
 	if err != nil {
 		return err
 	}
-	if err = c.Patch(ctx, before, patch); err != nil {
+	if err = c.Patch(ctx, objOriginal, patch); err != nil {
 		return err
 	}
 	return nil
 }
 
-func PatchStatus(ctx context.Context, c client.Client, before, after client.Object) error {
-	patch, err := CreatePatch(before, after)
+// PatchStatus applies the merge patch of client.Object status with SSA.
+// If strict is true, the resourceVersion will be part of the patch, make this call fail if
+// client.Object was changed.
+func PatchStatus(ctx context.Context, c client.Client, obj client.Object, strict bool, update func() (bool, error)) error {
+	objOriginal := obj.DeepCopyObject().(client.Object)
+	if strict {
+		objOriginal.SetResourceVersion("")
```

Comment:
Will this make the patch strict? Can you provide some reference or test to confirm? 
I know that in other places we achieve strict differently, like [here](https://github.com/kubernetes-sigs/kueue/blob/b0f6dd7951323dce577ba3699bec7915f0dce2e9/pkg/workload/workload.go#L512-L514).

Reply:
> Will this make the patch strict?

Yes sure. When we are setting `resourceVersion=""` and after that creating a patch we are comparing empty string `""` of `objOriginal` and some version `"243"` of `obj`. In result we will have `{"metadata":{"resourceVersion":"243"}}`

> I know that in other places we achieve strict differently, like [here](https://github.com/kubernetes-sigs/kueue/blob/b0f6dd7951323dce577ba3699bec7915f0dce2e9/pkg/workload/workload.go#L512-L514).

On this case the same. `BaseSSAWorkload` creates an workload with `resourceVersion=""`. Thats mean that `resourceVersion` will not add to the patch. But when we add version it will.

Reply:
Thanks for explaining, sgtm

Reply:
Maybe just add a comment like "clearing Resource version from the original object to make sure it is included in the generated patch"

Reply:
Since we were using Update before, it was strict, so I think we should keep it the same to avoid race conditions.

Reply:
You mean NOT strict? Because Update not return this error. 

```go
func (j *Job) Stop(ctx context.Context, c client.Client, podSetsInfo []podset.PodSetInfo, _ jobframework.StopReason, eventMsg string) (bool, error) {
	stoppedNow := false
	if !j.IsSuspended() {
		j.Suspend()
		if err := c.Update(ctx, j.Object()); err != nil {
			return false, fmt.Errorf("suspend: %w", err)
		}
		stoppedNow = true
	}

	// Reset start time if necessary, so we can update the scheduling directives.
	if j.Status.StartTime != nil {
		j.Status.StartTime = nil
		if err := c.Status().Update(ctx, j.Object()); err != nil {
			return stoppedNow, fmt.Errorf("reset status: %w", err)
		}
	}

	if changed := j.RestorePodSetsInfo(podSetsInfo); !changed {
		return stoppedNow, nil
	}
	if err := c.Update(ctx, j.Object()); err != nil {
		return false, fmt.Errorf("restore info: %w", err)
	}
	return stoppedNow, nil
}
```

Reply:
Ah, got your point. It updating job after successful update.